sympy/solvers/solveset.py

```diff
@@ -853,7 +853,7 @@ def solve_decomposition(f, symbol, domain):
     return y_s
 
 
-def _solveset(f, symbol, domain, _check=False):
+def _solveset(f, symbol, domain, flags={'tsolve_saw': []}, _check=False):
```

Comment:
Use of `flags` has been done to handle recursive calls. I am not sure if this is the best way to handle in `solveset`, but up till now, it seems to work well.

Reason for recursion:

- There are times where the equation is treated as exponent, gets converted to log and solveset is unable to handle, and passes it to transolve as log type which solves as logarithmic and consequently, this equation gets converted to the original equation which causes recursion.

Reply:
Perhaps this is a symptom of ill-defined `_is` routines which answer True when in fact they will not resolve with the associated solver.

Reply:
Just a thought will it be a good idea if we attempt by solving as lambert first and then when it raises `NotImplementedError` we can check if they are `exponential` or `logarithmic`, still the problem may persist but it will at least remove recursions of lambert type eqs that were caused previously.

Reply:
*gasp*...no. Maybe I am being penny-wise, but checking to see if it matches any of the lambert forms is going to punish any speed gains that you had for the much more common log and exp forms. Check for easy things first.

Reply:
Maybe we don't check if it is of lambert form, we directly get all the equations to its solver and get into exponential and logarithmic once `NotImplmentedError` is raised.

Reply:
> Perhaps this is a symptom of ill-defined _is routines which answer True when in fact they will not resolve with the associated solver.

Okay I think we need some amendments in `_is_exponential()`, it should be specifically made to identify two terms (power/exp) and both of them having symbol in their exponent

```python

def _is_exponential(f, symbol):
    assert len(f.args) == 2
    expr_args = f.args
    all(isinstance(expr_arg, (Pow, exp)) and (
            symbol in expr_arg.exp.free_symbols) for expr_arg in expr_args)
```
Because having one exponential type argument and other being non-expo-arg is more likely a lambert form like: `3**x - x**3`